### PR TITLE
Kevin Siess_ Abgabe_03_(resulting count: 1306)

### PR DESCRIPTION
**Pull request by JaeWoo-Westside**

### CODE REVIEW

File: Exercises/KevinSiess/KevinSiess_exercises_03.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+# - 123334 is not a valid number as there is no group of exactly two repeated digits
+# - 111334 is a valid number. while there are three 1s, there is also a group of exactly two 3s.
+# - 112233 is a valid number. At least one group of two is fulfilled, there is no maximum to the number of such groups.
+#
+# run your function with the lower bound `134564` and the upper bound `585159`. Note the resulting count
+# in your pull request, please.
+
+def count_valid_numbers(lower, upper):
+    # check for digit_twins
+    def has_exactly_two_adjacent_matching_digits(number):
```
Comment (PixelWelt) [inline Exercises/KevinSiess/KevinSiess_exercises_03.py:99]: what is the benefit of defining the function in the inner scope?

File: Exercises/KevinSiess/KevinSiess_exercises_03.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                count += 1
+            else:
+                if count == 2:  
+                    return True
+                count = 1  
+        return count == 2 
+    #check if digits increase from left to right
+    def digits_increase(number):
+        number_str = str(number)
+        for i in range(1, len(number_str)):
```
Comment (PixelWelt) [inline Exercises/KevinSiess/KevinSiess_exercises_03.py:113]: you could copy your number_str and sort it. After that you can just make one test if the two are equal

File: Exercises/KevinSiess/KevinSiess_exercises_03.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        if has_exactly_two_adjacent_matching_digits(number) and digits_increase(number):
+            valid_count += 1
+
+    return valid_count
+
+
+lower_bound = 134564
+upper_bound = 585159
+result = count_valid_numbers(lower_bound, upper_bound)
+print("Count of valid numbers:", result)
```
Comment (PixelWelt) [inline Exercises/KevinSiess/KevinSiess_exercises_03.py:129]: f strings could've been used here

**Review by PixelWelt — CHANGES_REQUESTED** [review]
please remove exercise 1 from PR
A french deck has 52 cards

All the other changes or ideas are optional and not needed for an approval

File: Exercises/KevinSiess/KevinSiess_exercises_03.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        self._cards = [Card(rank, suit) for suit in self.suits for rank in self.ranks]
+
+    def __len__(self):
+        return len(self._cards)
+
+    def __getitem__(self, position):
+        return self._cards[position]
+
+# Skatdeck
+class Skatdeck:
```
Comment (PixelWelt) [inline Exercises/KevinSiess/KevinSiess_exercises_03.py:44]: you could makr Skatdeck inherit from frenchdeck to write less code with the same function

File: Exercises/KevinSiess/KevinSiess_exercises_01.py
[the remark this answers is not in the document]
Comment (JaeWoo-Westside) [reply Exercises/KevinSiess/KevinSiess_exercises_01.py:1]: changed the name and the previous vile is in the wrong folder

File: Exercises/KevinSiess/KevinSiess_exercises_03.py [hunk 5]
```diff
@@ -41,40 +41,31 @@ def __getitem__(self, position):
...
 
 # Assertions for Frenchdeck
-assert len(french_deck) == 56, "Frenchdeck should have 56 cards"
+assert len(french_deck) == 52, "Frenchdeck should have 52 cards"
 assert french_deck[0].__repr__() == "2 of Diamonds", "The first card should be '2 of Diamonds'"
 assert french_deck[-1].__repr__() == "Ace of Clubs", "The last card should be 'Ace of Clubs'"
 
 # Instantiate and test Skatdeck
-print("\nTesting Skatdeck:")
+print(f"\nTesting Skatdeck:")
```
Comment (PixelWelt) [inline Exercises/KevinSiess/KevinSiess_exercises_03.py]: why are you adding f without printing a variable. in this case the f is obsolete

File: Exercises/KevinSiess/KevinSiess_exercises_03.py [hunk 6]
```diff
@@ -126,5 +123,5 @@ def digits_increase(number):
 lower_bound = 134564
 upper_bound = 585159
 result = count_valid_numbers(lower_bound, upper_bound)
-print("Count of valid numbers:", result)
+print(f"Count of valid numbers: {result}")
```
Comment (PixelWelt) [inline Exercises/KevinSiess/KevinSiess_exercises_03.py]: nice

**Review by PixelWelt — APPROVED** [review]
nice work

**Pull request merged**